app/http.php

```diff
@@ -36,11 +44,53 @@
         // 'document_root' => __DIR__.'/../public',
         // 'enable_static_handler' => true,
         'http_compression' => true,
+        'dispatch_func' => 'dispatch',
         'http_compression_level' => 6,
         'package_max_length' => $payloadSize,
         'buffer_output_size' => $payloadSize,
     ]);
 
+function dispatch(Server $server, $fd, $type, $data = null)
+{
+    global $workerNumber, $domains;
+
+    $safeThreadsPercent = intval(App::getEnv('_APP_SAFE_THREADS_PERCENT', 80)) / 100;
+
+    $safeThreads = (int) floor($workerNumber * $safeThreadsPercent);
+
+    $risky = false;
+    if ($data && str_contains($data, 'POST') && str_contains($data, '/executions')) {
+        $risky = true;
+    } elseif ($data && str_contains($data, '.appwrite.global')) {
+        $risky = true;
+    } elseif ($data) {
+        $lines = explode($data, '\n');
+        if (count($lines) > 2) {
+            $host = trim(explode('Host: ', $lines[1])[1]);
+            if (array_key_exists($host, $domains)) {
+                $risky = true;
+            }
+        }
+    }
+
+    if ($risky) {
+        var_dump('execution request sending to unsafe thread');
+        for ($j = $safeThreads; $j < $workerNumber; $j++) {
+            if ($server->getWorkerStatus($j) === 2) {
+                return $j;
+            }
+        }
+        return rand($safeThreads, $workerNumber - 1);
+    }
+
+    for ($i = 0; $i < $workerNumber; $i++) {
+        if ($server->getWorkerStatus($i) === 2) {
+            return $i;
+        }
+    }
+    return rand(0, $safeThreads - 1);
```

Comment:
Can we do `rand(0, workerNumber - 1)` here instead? If I understand correctly, this is for safe requests. We can use any worker for that.

Reply:
no because we reach here only hwne there are no idle threads, and we don't want safe request going to unsafe thread and get blocked in the loop like before.